apriltags_ros/src/apriltag_detector.cpp

```diff
@@ -340,32 +339,45 @@ void AprilTagDetector::imageCb(const sensor_msgs::PointCloud2ConstPtr& cloud,
           transformedPose.getRotation().getY(),
           transformedPose.getRotation().getZ(),
           transformedPose.getRotation().getW());
-        tf::poseTFToMsg(transformedPose, pose);
+        tf::poseTFToMsg(transformedPose, april_tag_raw_pose);
         tf::poseTFToMsg(transformedPlanePose, planePose);
       }
 
       geometry_msgs::PoseStamped tag_pose;
-      tag_pose.pose = pose;
+      tag_pose.pose = april_tag_raw_pose;
       tag_pose.header = header;
 
+      AprilTagDetection tag_detection;
+      tag_detection.pose = tag_pose;
+      tag_detection.id = detection.id;
+      tag_detection.size = tag_size;
+      
       if (validPose)
       {
-        AprilTagDetection tag_detection;
-        tag_detection.pose = tag_pose;
-        tag_detection.id = detection.id;
-        tag_detection.size = tag_size;
-
-        if (tracked_april_tags_.find(detection.id) == tracked_april_tags_.end()) {
-          auto sp = std::make_shared<DetectionPosesQueueWrapper>();
-            tracked_april_tags_[detection.id] = sp;
+        if (tracked_april_tags_with_valid_pose_.find(detection.id) == tracked_april_tags_with_valid_pose_.end()) {
+          tracked_april_tags_with_valid_pose_[detection.id] = std::make_shared<DetectionPosesQueueWrapper>();
         }
-        auto match = tracked_april_tags_[detection.id]; // we either have a match or we just created one
+        auto match = tracked_april_tags_with_valid_pose_[detection.id]; // we either have a match or we just created one
 
         match->posesQueue_.push_back(tag_detection);
         match->descriptionsQueue_.push_back(description);
         match->lastUpdated_ = std::chrono::steady_clock::now();
       }
 
+      // Publish all tags for camera validation tests
+      if (all_tracked_april_tags_.find(detection.id) == all_tracked_april_tags_.end()) {
+        all_tracked_april_tags_[detection.id] = std::make_shared<DetectionPosesQueueWrapper>();
+      }
+      auto match = all_tracked_april_tags_[detection.id]; // we either have a match or we just created one
+      if (match->posesQueue_.empty()) {
+        match->posesQueue_.push_back(tag_detection);
+        match->descriptionsQueue_.push_back(description);
+      } else {
+        match->posesQueue_[0]=tag_detection;
+        match->descriptionsQueue_[0]=description;
+      }
```

Comment:
is this just to have one pose for each april tag detected?

Reply:
yes. We care about the latest pose.

Reply:
cool, understand it now